File: companies-leetcode/google/arrays-and-strings/longest-substring-with-at-most-k-distinct-characters.js
/*
Given a string, find the length of the longest substring T that contains at most k distinct characters.

Example 1:

Input: s = "eceba", k = 2
Output: 3
Explanation: T is "ece" which its length is 3.
Example 2:

Input: s = "aa", k = 1
Output: 2
Explanation: T is "aa" which its length is 2.
*/
// 26th percentile
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var lengthOfLongestSubstringKDistinct = function(s, k) {
    let rightMost = {};
    let res = 0;
    let start = 0;
    for (let i = 0; i < s.length; i++) {
        let c = s[i];
        rightMost[c] = i;
        if (Object.keys(rightMost).length > k) {
            start = Math.min(...Object.values(rightMost))
            delete rightMost[s[start]];
            start += 1;
        }
        res = Math.max(res, i-start+1);
    }
    return res;
};


// 15th percentile
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var lengthOfLongestSubstringKDistinct = function(s, k) {
    let counts = {};
    let seen = new Set();
    let res = 0;
    let start = 0;
    for (let i = 0; i < s.length; i++) {
        let c = s[i];
        seen.add(c);
        if (Object.keys(counts).indexOf(c) !== -1) {
            counts[c] += 1
        } else {
            counts[c] = 1;
        }
        while (seen.size > k) {
            let r = s[start];
            counts[r] -= 1;
            if (counts[r] <= 0) {
                seen.delete(r);
            }
            start += 1;
        }
        res = Math.max(res, i-start+1);
    }
    
    return res;
};



// 99th percentile
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var lengthOfLongestSubstringKDistinct = function(s, k) {
    if(!s || !k) return 0
    
    let rightMost = new Map();
    let res = 0;
    let start = 0;
    for (let i = 0; i < s.length; i++) {
        let c = s[i];
        rightMost.set(c, i);
        
        if (rightMost.size > k) {
            let min = s.length,
                keyToRemove;
            
            for (let [key, value] of rightMost.entries()) {
                if (value < min) {
                    min = value;
                    keyToRemove = key;
                }
            }
            start = min + 1;
            rightMost.delete(keyToRemove);
        }
        res = Math.max(res, i-start+1);
    }
    return res;
};



// 23rd percentile
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var lengthOfLongestSubstringKDistinct = function(s, k) {
    if(!s || !k) return 0
    
    // let rightMost = new Map();
    let rightMost = {};
    let res = 0;
    let start = 0;
    for (let i = 0; i < s.length; i++) {
        let c = s[i];
        // rightMost.set(c, i);
        rightMost[c] = i;
        
        if (Object.keys(rightMost).length > k) {
            var min = s.length,
                keyToRemove;
            
            Object.keys(rightMost).forEach(key => {
                if (rightMost[[key]] < min) {
                    min = rightMost[[key]];
                    keyToRemove = key;
                }  
            })
            
            start = min + 1;
            delete rightMost[[keyToRemove]];
        }
        res = Math.max(res, i-start+1);
    }
    return res;
};

File: companies-leetcode/google/arrays-and-strings/longest-substring-with-at-most-k-distinct-characters.py
"""
Given a string, find the length of the longest substring T that contains at most k distinct characters.

Example 1:

Input: s = "eceba", k = 2
Output: 3
Explanation: T is "ece" which its length is 3.
Example 2:

Input: s = "aa", k = 1
Output: 2
Explanation: T is "aa" which its length is 2.
"""
# 68th percentile
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        if not s or not k: 
            return 0
        
        res = start = end = 0        
        # seen = defaultdict(lambda:0)
        seen = defaultdict(int)
        for i, c in enumerate(s):
            seen[c] += 1
            if len(seen) <= k:
                res = max(res, end-start+1)
            else:
                while len(seen) > k:
                    seen[s[start]] -= 1
                    if seen[s[start]]==0: 
                        del seen[s[start]]
                    start += 1
            end += 1
        return res

# 94th percentile
# We don't need the full counts. Just the rightmost values. 
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        if not s or not k: 
            return 0
        
        res = start = end = 0        
        seen = {}
        for i, c in enumerate(s):
            seen[c] = i
            if k < len(seen):
                start = min(seen.values())
                del seen[s[start]]
                start += 1
            res = max(res, i - start + 1)
        return res
                 
"""
Notes:

1)
The straightforward answer is to a leader/lagger approach. We advanced the leader
until we have seen too many unique characters. Then we advance the lagger until
we don't have too many characters. 

2)
We can do better. No matter what, we are going to advance the leader through every single
character in the line. The goal is to decrease the time it takes to update the lagger. 

consider a case where we have

cb...c..c..a
Assume that is the only b in the string.
where the periods could be anything. Assume we hit too many unique chars when we hit a.
At that point we want to advance the lagger. We can't just increase it to get rid of c, because
we actually want to increase it to just past b, the first char we can eliminate by advancing.
To do this, we keep a dictionary mapping chars to the last time they were seen. Then, we
advance the lagger (start in the code above) to be just past that element. 

"""